discovery_engine_core/core/src/storage/sqlite.rs

```diff
@@ -872,13 +901,16 @@ impl FeedbackScope for SqliteStorage {
     }
 
     async fn update_source_reaction(&self, source: &str, liked: bool) -> Result<(), Error> {
-        match self.fetch_source_reaction(source).await? {
-            None => self.create_source_reaction(source, liked).await,
-            Some(reaction) if reaction == liked => {
-                self.update_source_weight(source, liked.into()).await
+        match self.fetch_source_weight(source).await? {
+            0 => self.create_source_reaction(source, liked).await?,
+            weight if (weight > 0) == liked => {
+                if liked {
+                    self.update_source_weight(source, 1).await?;
+                }
             }
-            _ => self.delete_source_reaction(source).await,
+            _ => self.delete_source_reaction(source).await?,
```

Comment:
:thought_balloon: While this is the correct business logic I wonder if we should change it.

Currently it's:

- if your reactions differ from the previous reaction reset it to no reaction (0)
- if the reaction is positive add 1 (first reaction is 1)
- if the reaction is negative do nothing, except for the first reaction where it's set to -1.

Wouldn't it be more useful to also count negative reactions?

And it setting reaction to 0/delete once it differs once really the right approach?

Reply:
In the last commit I did unify `create_source_reaction` and `update_source_weight`.

Reply:
>if the reaction is negative do nothing, except for the first reaction where it's set to -1.

:wrench:  it's not quite exactly do nothing - `lastUpdated` should still be refreshed

>Wouldn't it be more useful to also count negative reactions?
>
>And it setting reaction to 0/delete once it differs once really the right approach? 

I don't see a problem with tracking the negative reactions but perhaps it's a bit out of scope of this pr and we should consider it separately? Similarly with deleting a "reset reaction" - I'd prefer to keep the logic as is here, and later double-check with ex-pink re changes to the logic. whether it's the _right_ approach - I had a similar question when first implementing it but that was the heuristic: see [here](https://xainag.atlassian.net/wiki/spaces/M2D/pages/2389639169/News+Source+Ranking#Updating-the-Source-Database%3A) 2nd point.

Reply:
>  it's not quite exactly do nothing - lastUpdated should still be refreshed

yes :+1: 

> Similarly with deleting a "reset reaction"

deleting on "reset" it's not new logic and its already in the code base before this PR and I think was this way in dart, too

but it does eliminate the `lastUpdated`, through we don't use the `lastUpdated` currently at all :thinking:

Reply:
@finiteprods

Reply:
>deleting on "reset" it's not new logic and its already in the code base

yes, correct - what I meant was we should keep this logic too, so all good.

>we don't use the `lastUpdated` currently at all

true. this was used in the python version to purge older sources but wasn't implemented here. I added a ticket [TO-3387](https://xainag.atlassian.net/browse/TO-3387) to reinvestigate this and other things mentioned above.